ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/KVOffset.scala

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) 2020 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils
+
+import com.daml.ledger.participant.state.v1.Offset
+
+object KVOffset {
+  private[kvutils] val highestStart = 0
+  private[kvutils] val middleStart = 8
+  private[kvutils] val lowestStart = 12
+  private[kvutils] val end = 16
+
+  private val maxValuePlusOne = BigInt(1) << (end * 8)
+
+  def onlyKeepHighestIndex(offset: Offset): Offset = {
+    val highest = highestIndex(offset)
+    fromLong(highest)
+  }
+
+  def setMiddleIndex(offset: Offset, middle: Int): Offset = {
+    val highest = highestIndex(offset)
+    val lowest = lowestIndex(offset)
+    fromLong(highest.toLong, middle, lowest.toInt)
+  }
+
+  def setLowestIndex(offset: Offset, lowest: Int): Offset = {
+    val highest = highestIndex(offset)
+    val middle = middleIndex(offset)
+    fromLong(highest.toLong, middle.toInt, lowest)
+  }
+
+  def fromLong(first: Long, second: Int = 0, third: Int = 0): Offset = {
+    val highest = BigInt(first) << ((middleStart - highestStart) * 8)
+    val middle = BigInt(second) << ((lowestStart - middleStart) * 8)
+    val lowest = BigInt(third)
```

Comment:
These bit-shifts only work by accident. If, for example, `middleStart` was `9` and `lowestStart` was `13`, they wouldn't work: you'd shift `highest` by `9` when it should be `7`, and `middle` by `4` when it should be `3`.

You need to use `end - middleStart` and `end - lowestStart` (and implicitly `end - end` for `lowest`, which is unnecessary because it's always `0`).

Reply:
good catch. thanks. I shouldn't have trusted those "nice" numbers to just work out.